packages/checkout-storefront/src/state/updateStateStore.ts

```diff
@@ -0,0 +1,107 @@
+import create from "zustand";
+import { CheckoutScope } from "@/checkout-storefront/hooks/useAlerts";
+import shallow from "zustand/shallow";
+import { useMemo } from "react";
+import { memoize, omit } from "lodash-es";
+
+export type CheckoutUpdateStateStatus = "success" | "loading" | "error";
+
+export type CheckoutUpdateStateScope = Extract<
+  CheckoutScope,
+  | "checkoutShippingUpdate"
+  | "checkoutCustomerAttach"
+  | "checkoutAddPromoCode"
+  | "checkoutDeliveryMethodUpdate"
+  | "checkoutEmailUpdate"
+  | "checkoutBillingUpdate"
+  | "checkoutLinesUpdate"
+  | "userRegister"
+>;
+
+interface CheckoutUpdateStateStore {
+  shouldRegisterUser: boolean;
+  loadingCheckout: boolean;
+  updateState: Record<CheckoutUpdateStateScope, CheckoutUpdateStateStatus>;
+  actions: {
+    setShouldRegisterUser: (shouldRegisterUser: boolean) => void;
+    setLoadingCheckout: (loading: boolean) => void;
+    setUpdateState: (
+      scope: CheckoutUpdateStateScope
+    ) => (status: CheckoutUpdateStateStatus) => void;
+  };
+}
+
+const useCheckoutUpdateStateStore = create<CheckoutUpdateStateStore>((set) => ({
+  shouldRegisterUser: false,
+  loadingCheckout: false,
+  updateState: {
+    checkoutShippingUpdate: "success",
+    checkoutCustomerAttach: "success",
+    checkoutBillingUpdate: "success",
+    checkoutAddPromoCode: "success",
+    checkoutDeliveryMethodUpdate: "success",
+    checkoutLinesUpdate: "success",
+    checkoutEmailUpdate: "success",
+    userRegister: "success",
+  },
+  actions: {
+    setShouldRegisterUser: (shouldRegisterUser: boolean) =>
+      set({
+        shouldRegisterUser,
+      }),
+    setLoadingCheckout: (loading: boolean) => set({ loadingCheckout: loading }),
+    setUpdateState: memoize(
+      (scope) => (status) =>
+        set((state) => ({
+          updateState: {
+            ...state.updateState,
+            [scope]: status,
+          },
+          // checkout will reload right after, this ensures there
+          // are no rerenders in between where there's no state updating
+          // also we might not need this once we get better caching
+          loadingCheckout: status === "success" || state.loadingCheckout,
+        }))
+    ),
+  },
+}));
+
+export const useCheckoutUpdateState = () => {
+  const { updateState, loadingCheckout } = useCheckoutUpdateStateStore(
+    ({ updateState, loadingCheckout }) => ({
+      updateState,
+      loadingCheckout,
+    }),
+    shallow
+  );
+
+  return useMemo(
+    () => ({ updateState, loadingCheckout }),
+    // because we want to compare array of strings instead of object
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [loadingCheckout, ...Object.values(updateState)]
```

Comment:
Why do we want to compare an array of strings? Won't it depend on the order in which values are updated in the store, i.e.
`{ checkoutShippingUpdate: "success", checkoutCustomerAttach: "error" }`
`{ checkoutCustomerAttach: "error", checkoutShippingUpdate: "success" }`
will be treated as not equal?